Lib/functools.py

```diff
@@ -273,33 +273,101 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
+
+class __PlaceholderTypeBase:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
+    """
+    __instance = None
+
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
+
+    def __bool__(self):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+
+Placeholder = type('PlaceholderType', (__PlaceholderTypeBase,), {})()
+
+
 # Purely functional, no descriptor behaviour
 class partial:
     """New function with partial application of the given arguments
     and keywords.
     """
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    __slots__ = ("func", "args", "keywords", "placeholder_count",
+                 "__dict__", "__weakref__")
```

Comment:
```suggestion
    __slots__ = ("func", "args", "keywords", "placeholder_count",
                 "__dict__", "__weakref__")

    Placeholder = Placeholder
```

Adding a class variable would be more convenient for member access.

Reply:
I guess it would be "nicer", but much slower.

Getting global `Placeholder` is fastest.

Retrieving instance attribute is a bit slower. 

Retrieving class attribute is slower than the 2 above.

Reply:
> Adding a class variable would be more convenient for member access.

Sorry for the unclear expression. It's not about performance but usability. I mean the users can have a single import if we assign the sentinal variable to class `partial`.

```python
from functools import partial

partial_func = partial(func, partial.Placeholder, ...)
```

vs.

```python
from functools import partial, Placeholder

partial_func = partial(func, Placeholder, ...)
```

which has less code changes.

Also, this namespaced `Placeholder` variable is more readable since it is only used by `parital`.

Reply:
Ah ok. This is a good idea, I have thought about it too. 

The issue I had was that I couldn't find an "appropriate" way to do the same for C class.

Also, personally, I wouldn't use the first example. I wouldn't use the 2nd either.

My way of doing things is keeping everything in their namespaces without `from` imports:

```python
import functools as ftl
ftl.partial(func, ftl.Placeholder, ...)
```

However, if this is something that is deemed to be needed and you can find me a good way to add it to C class, I can do it.

https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dict

"Warning It is not safe to use [PyDict_SetItem()](https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItem) on or otherwise modify [tp_dict](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dict) with the dictionary C-API."

Reply:
I don't see the benefit of adding a field and would rather minimize the layout of a class if it's only for usability. Personally, if we were to use a placeholder *in* the class, then shortening its name would make more sense because having `partial.Placeholder` feels way too long.

While there is some precedence in `Signature.empty`, this field actually *exposes* a private name, so it is fine. But for public names, we don't have that kind of logic.

Reply:
For `partial` I always redefine anyways:
```python
PH = functools.Placeholder
```

And `inspect` is pure python module, so no need to keep 2 implementations in sync. This case is a bit different. Adding complexity here is more costly.

Reply:
I know that it's pure python but I wanted to highlight the semantical context. The fact that we do not duplicate constants when it's not needed should be independent of whether there is a C implementation or not, it is more about the consistency of its usage IMO.